examples/cpp/dds/CustomPayloadPoolExample/CustomPayloadPoolDataPublisher.cpp

```diff
@@ -0,0 +1,210 @@
+{
+    hello_.index(0);
+    hello_.message("CustomPayloadPool");
+    DomainParticipantQos pqos = PARTICIPANT_QOS_DEFAULT;
+    pqos.name("Participant_pub");
+    auto factory = DomainParticipantFactory::get_instance();
+
+    participant_ = factory->create_participant(0, pqos);
+
+    if (participant_ == nullptr)
+    {
+        return false;
+    }
+
+    //REGISTER THE TYPE
+    type_.register_type(participant_);
+
+    //CREATE THE PUBLISHER
+    PublisherQos pubqos = PUBLISHER_QOS_DEFAULT;
+
+    publisher_ = participant_->create_publisher(
+        pubqos,
+        nullptr);
+
+    if (publisher_ == nullptr)
+    {
+        return false;
+    }
+
+    //CREATE THE TOPIC
+    TopicQos tqos = TOPIC_QOS_DEFAULT;
+
+    topic_ = participant_->create_topic(
+        "CustomPayloadPoolTopic",
+        "CustomPayloadPoolData",
+        tqos);
+
+    if (topic_ == nullptr)
+    {
+        return false;
+    }
+
+    // CREATE THE WRITER
+    DataWriterQos wqos = DATAWRITER_QOS_DEFAULT;
+
+    writer_ = publisher_->create_datawriter(
+        topic_,
+        wqos,
+        this,
+        StatusMask::all(),
+        payload_pool_);
+
+    if (writer_ == nullptr)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+CustomPayloadPoolDataPublisher::~CustomPayloadPoolDataPublisher()
+{
+    if (writer_ != nullptr)
+    {
+        publisher_->delete_datawriter(writer_);
+    }
+    if (publisher_ != nullptr)
+    {
+        participant_->delete_publisher(publisher_);
+    }
+    if (topic_ != nullptr)
+    {
+        participant_->delete_topic(topic_);
+    }
+    DomainParticipantFactory::get_instance()->delete_participant(participant_);
+}
+
+void CustomPayloadPoolDataPublisher::on_publication_matched(
+        eprosima::fastdds::dds::DataWriter*,
+        const eprosima::fastdds::dds::PublicationMatchedStatus& info)
+{
+    if (info.current_count_change == 1)
+    {
+        matched_ = info.total_count;
+        first_connected_ = true;
+        std::cout << "Publisher matched." << std::endl;
+    }
+    else if (info.current_count_change == -1)
+    {
+        matched_ = info.total_count;
+        std::cout << "Publisher unmatched." << std::endl;
+    }
+    else
+    {
+        std::cout << info.current_count_change
+                  << " is not a valid value for PublicationMatchedStatus current count change" << std::endl;
+    }
+}
+
+void CustomPayloadPoolDataPublisher::run_thread(
+        uint32_t samples,
+        uint32_t sleep)
+{
+    if (samples == 0)
+    {
+        while (!stop_)
+        {
+            if (publish(false))
+            {
+                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
+                          << " SENT" << std::endl;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+        }
+    }
+    else
+    {
+        for (uint32_t i = 0; i < samples; ++i)
+        {
+            if (!publish())
```

Comment:
Consider using a CV here for the `wait_for_listener=true` case, else we have a busy waiting